src/Data/Parameterized/Vector.hs

```diff
@@ -0,0 +1,382 @@
+{-# Language GADTs, DataKinds, TypeOperators, BangPatterns #-}
+{-# Language TypeApplications, ScopedTypeVariables #-}
+{-# Language Rank2Types #-}
+-- | A vector fixed-size vector of typed elements.
+module Data.Parameterized.Vector
+  ( Vector
+    -- * Lists
+  , fromList
+  , toList
+
+    -- * Length
+  , length
+  , nonEmpty
+  , lengthInt
+
+  -- * Indexing
+  , elemAt
+  , elemAtMaybe
+  , elemAtUnsafe
+
+  -- * Update
+  , insertAt
+  , insertAtMaybe
+
+    -- * Sub sequences
+  , uncons
+  , slice
+  , Data.Parameterized.Vector.take
+
+    -- * Zipping
+  , zipWith
+  , zipWithM
+  , zipWithM_
+  , interleave
+
+    -- * Reorder
+  , shuffle
+  , rotateL
+  , rotateR
+  , shiftL
+  , shiftR
+
+    -- * Splitting and joining
+    -- ** General
+  , joinWith
+  , splitWith
+
+    -- ** Vectors
+  , split
+  , join
+  , append
+
+  ) where
+
+import qualified Data.Vector as Vector
+import Data.Vector.Mutable (MVector)
+import qualified Data.Vector.Mutable as MVector
+import Control.Monad.ST
+import Data.Parameterized.NatRepr
+import Data.Proxy
+import Unsafe.Coerce
+import Data.Coerce
+import Prelude hiding (length,zipWith)
+
+import Data.Parameterized.Utils.Endian
+
+-- | Fixed-size non-empty vectors.
+data Vector n a where
+  Vector :: (1 <= n) => !(Vector.Vector a) -> Vector n a
+
+-- | Get the elements of the vector as a list, lowest index first.
+toList :: Vector n a -> [a]
+toList (Vector v) = Vector.toList v
+{-# Inline toList #-}
+
+-- | Length of the vector.
+-- @O(1)@
+length :: Vector n a -> NatRepr n
+length (Vector xs) = unsafeCoerce (fromIntegral (Vector.length xs) :: Integer)
```

Comment:
`unsafeCoerce` always gives me heebie-jeebies.  Can we weaken this do `Data.Coerce.coerce`?

Regardless, lets add some comments to this file about why we believe this cast is OK (basically, an invariant that all operations preserve the connection between the type index and the length of the vector.)

Reply:
We don't have to do any coercion at all if we expose the constructor `NatRepr`, but I get a link-time error when I do that. What's going on there?

Reply:
After a long discussion on Mattermost, I'm leaving this as-is for now. If we want to change this, let's do it in a separate PR.